cli/tests/integration/mod.rs

```diff
@@ -1202,26 +1206,29 @@ async fn listen_tls_alpn_fail() {
       let msg = std::str::from_utf8(&buffer).unwrap();
       assert_eq!(msg, "READY");
 
-      let mut cfg = rustls::ClientConfig::new();
-      let reader = &mut BufReader::new(Cursor::new(include_bytes!(
+      let mut reader = &mut BufReader::new(Cursor::new(include_bytes!(
         "../testdata/tls/RootCA.crt"
       )));
-      cfg.root_store.add_pem_file(reader).unwrap();
+      let certs = rustls_pemfile::certs(&mut reader).unwrap();
+      let mut root_store = rustls::RootCertStore::empty();
+      root_store.add_parsable_certificates(&certs);
+      let mut cfg = rustls::ClientConfig::builder()
+        .with_safe_defaults()
+        .with_root_certificates(root_store)
+        .with_no_client_auth();
       cfg.alpn_protocols.push("boofar".as_bytes().to_vec());
       let cfg = Arc::new(cfg);
 
-      let hostname =
-        webpki::DNSNameRef::try_from_ascii_str("localhost").unwrap();
-
+      let hostname = rustls::ServerName::try_from("localhost").unwrap();
       let tcp_stream = tokio::net::TcpStream::connect("localhost:4505")
         .await
         .unwrap();
       let mut tls_stream =
-        TlsStream::new_client_side(tcp_stream, &cfg, hostname);
+        TlsStream::new_client_side(tcp_stream, cfg, hostname);
       tls_stream.handshake().await.unwrap();
```

Comment:
I'm still debugging what's the problem here - when reading from the socket there's a single errors raised (`WouldBlock`) that is ignored and then socket immediately returns 0 bytes read (meaning stream was closed of EOF was reached).

Reply:
I run `cli/tests/testdata/listen_tls_alpn.ts` manually and sent a small message after establishing a connection. Instead of `tls handshake eof` I'm not getting:
```
thread 'integration::listen_tls_alpn_fail' panicked at 'called `Result::unwrap()` on an `Err` value: Custom { kind: InvalidData, error: AlertReceived(NoApplicationProtocol) }', cli/tests/integration/mod.rs:1228:36
```
It appears to me `rustls` changed behavior of handshaking and our logic doesn't reflect that.

Still debugging...

Reply:
It's a timing issue. The server that this client connects to is implemented as follows:

```ts
for await (const conn of listener) {
  conn.close();
}
```

Basically it accepts the connection and hangs up immediately, before the handshake is complete.
If I change that to:

```ts
for await (const conn of listener) {
  await conn.handshake();
  conn.close();
}
```

... it gets a little further, however now the test fails with:

```
Uncaught (in promise) InvalidData: peer doesn't support any known protocol
  await conn.handshake();
  ^
    at async file:///Users/piscisaureus/d/deno/cli/tests/testdata/listen_tls_alpn.ts:11:3
```

So it seems that the `alpnProtocols: ["h2", "http/1.1", "foobar"]` is not propagated through the stack correctly.

Reply:
But this test adds only a single ALPN: `boofar` which indeed is not recognized by the server side. So to me it seems like the error you see is correct